spec/controllers/profile_controller_spec.rb

```diff
@@ -22,7 +24,66 @@
+        @user = FactoryBot.create(:user_manager_user)
+        allow(request.env['warden']).to receive(:authenticate!).and_return(@user)
+        allow(controller).to receive(:current_user).and_return(@user)
+        sign_in @user
+        post :change_password, params: { password_change: { 'new_password1': 'wrong', 'new_password2': 'right' } }
+        expect(subject).to redirect_to(profile_path)
+        expect(flash.now[:error]).not_to be_nil
+        expect(flash.now[:error]).to eq(t('profile.password_mismatch'))
+      end
+
+      it 'advises password changed when successful' do
+        @request.env["devise.mapping"] = Devise.mappings[:user]
+        @user = FactoryBot.create(:user_manager_user)
+        allow(request.env['warden']).to receive(:authenticate!).and_return(@user)
+        allow(controller).to receive(:current_user).and_return(@user)
+        sign_in @user
+        post :change_password, params: { password_change: { 'old_password': 'oldPassword1', 'new_password1': 'newPassword1', 'new_password2': 'newPassword1' } }
+        expect(subject).to redirect_to(profile_path)
+        expect(flash.now[:error]).to be_nil
+        expect(flash.now[:notice]).not_to be_nil
+        expect(flash.now[:notice]).to eq(t('profile.password_changed'))
+      end
+
+      it 'errors when password does not meet acceptence criteria' do
+        @request.env["devise.mapping"] = Devise.mappings[:user]
+        @user = FactoryBot.create(:user_manager_user)
+        allow(request.env['warden']).to receive(:authenticate!).and_return(@user)
+        allow(controller).to receive(:current_user).and_return(@user)
+        sign_in @user
+        stub_cognito_response(
+          method: :change_password,
+          payload: "InvalidPasswordException"
+        )
+        post :change_password, params: { password_change: { 'old_password': 'oldPassword1', 'new_password1': 'newpassword', 'new_password2': 'newpassword' } }
+        expect(subject).to redirect_to(profile_path)
+        expect(flash.now[:error]).not_to be_nil
+        expect(flash.now[:error]).to eq(t('devise.sessions.InvalidPasswordException'))
+      end
+
+      it 'errors when old password is wrong' do
+        @request.env["devise.mapping"] = Devise.mappings[:user]
+        @user = FactoryBot.create(:user_manager_user)
+        allow(request.env['warden']).to receive(:authenticate!).and_return(@user)
+        allow(controller).to receive(:current_user).and_return(@user)
+        sign_in @user
+        stub_cognito_response(
+          method: :change_password,
+          payload: "NotAuthorizedException"
+        )
+        post :change_password, params: { password_change: { 'old_password': 'wrong_password', 'new_password1': 'newpassword', 'new_password2': 'newpassword' } }
+        expect(subject).to redirect_to(profile_path)
+        expect(flash.now[:error]).not_to be_nil
+        expect(flash.now[:error]).to eq(t('profile.old_password_mismatch'))
+      end
+    end
+
```

Comment:
what about a test when the old password is missing and you supply new ones correctly? the code would currently allow that and hit cognito with nil on the old password...

Reply:
But the missing password would be rejected by the change password request in cognito but I can catch that earlier and return a different flash message.

Reply:
yes it will be, but that's not a good practice... we shouldn't be hitting surrounding services with incomplete, malformed requests....